python/ray/tune/trial_runner.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import ray
+import time
+import traceback
+
+from ray.tune.trial import Trial
+
+
+class TrialRunner(object):
+    """A TrialRunner implements the event loop for scheduling trials on Ray.
+
+    Example:
+        runner = TrialRunner()
+        runner.add_trial(Trial(...))
+        runner.add_trial(Trial(...))
+        while not runner.is_finished():
+            runner.step()
+            print(runner.debug_string())
+
+    The main job of TrialRunner is scheduling trials to efficiently use cluster
+    resources, without overloading the cluster.
+
+    While Ray itself provides resource management for tasks and actors, this is
+    not sufficient when scheduling trials that may instantiate multiple actors.
+    This is because if insufficient resources are available, concurrent agents
+    could deadlock waiting for new resources to become available. Furthermore,
+    oversubscribing the cluster could degrade training performance, leading to
+    misleading benchmark results.
+    """
+
+    def __init__(self):
+        """Initializes a new TrialRunner."""
+
+        self._trials = []
+        self._pending = {}
+        self._avail_resources = {'cpu': 0, 'gpu': 0}
```

Comment:
This dict format appears very often; it _could_ make sense in readability and otherwise to make this into a named tuple "Resources"...